### ISSUE TITLE
No support of SCRAM authentification in PostgreSQL

### ISSUE THREAD

**Issue by signedav** [opengisch/QgisModelBaker#779]
See this issue https://github.com/claeis/ili2db/issues/448

**At least in the Model Baker Documentation should be an entry of this and a solution.**

The [documentation](https://www.postgresql.org/docs/current/auth-password.html) says (of course `scram-sha-256` could be replaced with `md5` to do the other way around):
> [..] set `password_encryption = 'scram-sha-256'` in `postgresql.conf`, make all users set new passwords, and change the authentication method specifications in `pg_hba.conf` to `scram-sha-256`.

What only a change in one of those files does?

**I think - but not sure - `postgresql.conf` defines how the password should be encrypted (hashed) when creating / alter a user and `pg_hba.conf` defines what kind of encryption (hash) is expected by the password sent by the client.**

But as mentioned in the first quote "[...] `postgresql.conf`, make all users set new passwords [...]" - so this step with `ALTER USER` is requested automatically. Not sure if this could be a little risky that currently running clients do not work anymore, when the encryption on server side changed.

Btw. I see in the documentation as well:
> If a password was encrypted using the `scram-sha-256` setting, then it can be used for the authentication methods `scram-sha-256` and password (but password transmission will be in plain text in the latter case). The authentication method specification `md5` will automatically switch to using the `scram-sha-256` method in this case, as explained above, so it will also work.

So it's possible that one changed it on `pg_hba.conf` but still the "previous" encryption is used...

**And here some additional info how to check the password encryption (and change it):**

To see how a password is encrypted we can check `pg_authid`:
```sql
> SELECT rolpassword from pg_authid where rolname = '{your-db-user-name}';
```

When we want to change it (e.g. to `md5`) we need to set the `password_encryption` and then alter the user:
```sql
> SHOW password_encryption;
scram-sha-256
-- not sure if this is set by the setting in `postgresql.conf`
> SET password_encryption  = 'md5';
> ALTER USER "{your-db-user-name}" with password '{your-db-password}';
```

**And where are the files?**

```sql
> SHOW hba_file;
              hba_file               
-------------------------------------
 /etc/postgresql/14/main/pg_hba.conf
(1 row)

> SHOW config_file;
               config_file               
-----------------------------------------
 /etc/postgresql/14/main/postgresql.conf
(1 row)
```

**Comment by signedav** [comment]
## Fix Steps

### 1. Change your password encryption

In SQL Console (e.g. PgAdmin):

```sql
> SHOW password_encryption;
scram-sha-256
-- not sure if this is set by the setting in `postgresql.conf`
> SET password_encryption  = 'md5';
> ALTER USER "{your-db-user-name}" with password '{your-db-password}';
```

#### 2. Change enryption in `postgresql.conf`

In SQL Console (e.g. PgAdmin):

```sql
> SHOW config_file;
               config_file               
-----------------------------------------
 /etc/postgresql/14/main/postgresql.conf
 ```

Und im `postgresql.conf` ändern zu `'md5'`:
```INI
password_encryption = 'scram-sha-256'
```

#### 2. Change enryption in `pg_hba.conf`

In SQL Console (e.g. PgAdmin):

```sql
> SHOW hba_file;
              hba_file               
-------------------------------------
 /etc/postgresql/14/main/pg_hba.conf
```

Und im `pg_hba.conf` ändere `'scram-sha-256'` zu `'md5'`:

**Comment by lacardonap** [comment]
@signedav thank you very much for the contribution, I followed the steps and everything worked fine.

**The error: No support of SCRAM authentification in PostgreSQL**

![Screenshot from 2023-01-28 12-53-06](https://user-images.githubusercontent.com/8827336/215284628-c8f312c4-e2f3-4552-ac34-3c45294f4358.png)

**Valid postgres version we are using:**
```sql
select version()

> PostgreSQL 15.1 (Ubuntu 15.1-1.pgdg22.04+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0, 64-bit
```

**Change enryption in postgresql.conf**

```sql
SHOW config_file;
/etc/postgresql/15/main/postgresql.conf
```
![Screenshot from 2023-01-28 12-56-55](https://user-images.githubusercontent.com/8827336/215284615-e06af2b6-47d6-46b0-9434-24c416950b86.png)

**Change enryption in pg_hba.conf**

```sql
SHOW hba_file;
/etc/postgresql/15/main/pg_hba.conf
```
![Screenshot from 2023-01-28 13-03-22](https://user-images.githubusercontent.com/8827336/215284610-41b9632e-fc10-42a6-95db-33824358363b.png)

**Restart postgres service to apply changes**

```bash
sudo systemctl restart postgresql.service
```

**Update password to md5**

```sql
SHOW password_encryption;
ALTER USER "postgres" with password 'secret';
SELECT rolpassword from pg_authid where rolname = 'postgres';
```

**Comment by signedav** [comment]
I think (not sure) pgjpdbc is already supporting it but ili2db is not yet using it (see https://github.com/claeis/ili2db/issues/448) since java "42.2.18.jre6" is used (targeting JRE 6). I don't know about the effort / risk to bump ili2db to a newer JRE (see it's marked as 5.0.0 milestone).

**Comment by andreasneumann** [comment]
@signedav - thank you for the clarification!

**Comment by signedav** [comment]
This is fixed with ili2db 5.0.0 :tada:
